### PR TITLE
[minor] Set header

### PR DESCRIPTION
**Pull request by ssunorz**
# Description

I used this as a reference.
https://github.com/golang/oauth2/blob/master/transport.go

## Type of change

- [ ] Bug fix
- [x] New feature
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Non-code changes (update documentation, pipeline, etc.)

### Flags

- [ ] breaks backward compatibility
- [x] requires a documentation update
- [ ] has untestable code

## Related issue/PR

**Delete if not applicable**
- Fixes #_issue_
- Closes #_PR_

---

## Checklist

- [x] Followed the guidelines in the CONTRIBUTING document
- [x] Added prefix `[major]`/`[minor]`/`[patch]`/`[skip]` in the PR title
- [x] Tested and linted the code
- [x] Commented the code
- [x] Made corresponding changes to the documentation
- [x] Confirmed no dropping in test coverage (by [Codecov](https://codecov.io/gh/yahoojapan/authorization-proxy/pulls))
- [x] Passed all pipeline checking
- [x] Approved by >1 reviewer

## Checklist for maintainer
- [x] Use `Squash and merge`
- [x] Double-confirm the merge message has prefix `[major]`/`[minor]`/`[patch]`/`[skip]`
- [x] Delete the branch after merge

### CODE REVIEW

File: handler/handler_test.go [hunk 1]
```diff
@@ -20,6 +26,33 @@ import (
 )
 
 func TestNew(t *testing.T) {
+	rt := &role.Token{
```
Comment (WindzCUHK) [inline handler/handler_test.go:38]: maybe it is better to just mock the interface than use the struct directly?

something like this:
https://github.com/yahoojapan/authorization-proxy/blob/c80098034a01d3b0ffc3c51736b6da624101b3b4/handler/roundtripper_mock.go

File: handler/transport_test.go [hunk 2]
```diff
@@ -56,8 +58,8 @@ func Test_transport_RoundTrip(t *testing.T) {
 					},
 				},
 				prov: &service.AuthorizerdMock{
-					VerifyFunc: func(r *http.Request, act, res string) error {
-						return nil
+					VerifyFunc: func(r *http.Request, act, res string) (authorizerd.Principal, error) {
+						return &role.Token{}, nil
```
Comment (WindzCUHK) [inline handler/transport_test.go]: use mock?

File: handler/handler_test.go (hunk 1 above)
Comment (ssunorz) [reply handler/handler_test.go]: Fixed
https://github.com/yahoojapan/authorization-proxy/pull/66/commits/72b90da34685ad95fb5ba452f94fd4ed26cb88a2

File: handler/transport_test.go (hunk 2 above)
Comment (ssunorz) [reply handler/transport_test.go]: Fixed
https://github.com/yahoojapan/authorization-proxy/pull/66/commits/72b90da34685ad95fb5ba452f94fd4ed26cb88a2

File: README.md [hunk 3]
```diff
@@ -85,6 +86,19 @@ The mapping rules are described as below.
 
 ⚠️ All the HTTP/HTTPS methods and URI paths are normalized to lower case.
 
+### HTTP request headers
+
+When a request is authorized by the authorization proxy, The following HTTP headers is added in the request.
```
Comment (WindzCUHK) [inline README.md:91]:
```suggestion
When a request is authorized by the authorization proxy, the following HTTP headers is added in the request.
```

File: README.md [hunk 4]
```diff
@@ -85,6 +86,19 @@ The mapping rules are described as below.
...
+
+| HTTP Header Name    | Description                                                              | Example           |
+|---------------------|--------------------------------------------------------------------------|-------------------|
+| X-Athenz-Principal  | Authorized principal                                                     | principal         |
+| X-Athenz-Role       | Authorized role (A comma-separated string if there is more than one)      | role1,role2,role3 |
+| X-Athenz-Domain     | Authorized domain                                                        | domain            |
+| X-Athenz-Client-ID  | Authorized client_id                                                     | client-id         |
```
Comment (WindzCUHK) [inline README.md:98]:
```suggestion
| X-Athenz-Client-ID  | Authorized client ID                                                     | client-id         |
```

File: handler/principal_mock.go [hunk 5]
```diff
@@ -0,0 +1,71 @@
+package handler
+
+import (
+	"github.com/yahoojapan/athenz-authorizer/v4/access"
+	"github.com/yahoojapan/athenz-authorizer/v4/role"
+)
+
+// RoleTokenMock is a mock of Principal
+type RoleTokenMock struct {
+	t role.Token
```
Comment (WindzCUHK) [inline handler/principal_mock.go:10]:
```suggestion
	t role.Token
```
Can we just implement the interface? not depending the struct in `athenz-authorizer`.
Same for `OAuth2AccessTokenMock`

File: handler/handler_test.go [hunk 6]
```diff
@@ -43,7 +89,7 @@ func TestNew(t *testing.T) {
 			srv := httptest.NewServer(handler)
 
 			return test{
-				name: "check request can redirect",
+				name: "Check that the request with roletoken headers is redirected",
```
Comment (WindzCUHK) [inline handler/handler_test.go:92]:
```suggestion
				name: "Check that the request with role token headers is redirected",
```

File: handler/handler_test.go [hunk 7]
```diff
@@ -66,9 +112,133 @@ func TestNew(t *testing.T) {
...
+				if err2 != nil {
+					w.WriteHeader(http.StatusNotImplemented)
+					return
+				}
+				w.WriteHeader(http.StatusOK)
+			}))
+			srv := httptest.NewServer(handler)
+
+			return test{
+				name: "Check that the request with accesstoken headers is redirected",
```
Comment (WindzCUHK) [inline handler/handler_test.go:183]:
```suggestion
				name: "Check that the request with access token headers is redirected",
```

File: handler/transport.go [hunk 8]
```diff
@@ -33,6 +36,8 @@ type transport struct {
 	cfg  config.Proxy
 }
 
+// Based on the following.
+// https://github.com/golang/oauth2/blob/master/transport.go
```
Comment (WindzCUHK) [inline handler/transport.go:40]:
```suggestion
// https://github.com/golang/oauth2/blob/bf48bf16ab8d622ce64ec6ce98d2c98f916b6303/transport.go
```
use permanent link

File: handler/principal_mock.go (hunk 5 above)
Comment (ssunorz) [reply handler/principal_mock.go]: Fixed
https://github.com/yahoojapan/authorization-proxy/pull/66/commits/2c72919512b30c5cdb8ef6709af43648d6fe52fe

File: handler/handler_test.go [hunk 9]
```diff
@@ -30,11 +36,51 @@ func TestNew(t *testing.T) {
...
+	at := access.OAuth2AccessTokenClaim{
+		AuthTime:       0,
+		Version:        0,
+		ClientID:       "client_id",
+		UserID:         "",
+		ProxyPrincipal: "",
+		Scope:          []string{"at_role1", "at_role2", "at_role3"},
+		Confirm:        nil,
+		BaseClaim:      bc,
+	}
```
Comment (WindzCUHK) [inline handler/handler_test.go:64]: Can we use mock here other than strucht?

Comment (ssunorz) [reply handler/handler_test.go]: Fixed
https://github.com/yahoojapan/authorization-proxy/pull/66/commits/879805e0405252b7f0189ce6e109502390404c5b

File: handler/handler_test.go [hunk 10]
```diff
@@ -139,23 +132,23 @@ func TestNew(t *testing.T) {
...
 					if err := f(key, want); err != nil {
 						return err
 					}
-					key, want = "X-Athenz-Domain", rt.Domain
+					key, want = "X-Athenz-Domain", "rt_domain"
 					if err := f(key, want); err != nil {
 						return err
 					}
-					key, want = "X-Athenz-Issued-At", strconv.FormatInt(rt.TimeStamp.Unix(), 10)
+					key, want = "X-Athenz-Issued-At", strconv.FormatInt(1595908257, 10)
```
Comment (WindzCUHK) [inline handler/handler_test.go:147]:
```suggestion
					key, want = "X-Athenz-Issued-At", "1595908257"
```
please also fix other `strconv.FormatInt()`

Comment (ssunorz) [reply handler/handler_test.go]: Fixed
https://github.com/yahoojapan/authorization-proxy/pull/66/commits/4459dce9813b711d85e81e3b494dc339d4c371f2

File: handler/transport.go [hunk 11]
```diff
@@ -42,10 +47,48 @@ func (t *transport) RoundTrip(r *http.Request) (*http.Response, error) {
...
 	}
 
-	if err := t.prov.Verify(r, r.Method, r.URL.Path); err != nil {
+	reqBodyClosed := false
+	if r.Body != nil {
+		defer func() {
+			if !reqBodyClosed {
+				r.Body.Close()
+			}
+		}()
```
Comment (WindzCUHK) [inline handler/transport.go:56]: Please also add unit test for this block. I think we can reference to the original code.
https://github.com/golang/oauth2/blob/bf48bf16ab8d622ce64ec6ce98d2c98f916b6303/transport_test.go

Comment (ssunorz) [reply handler/transport.go:56]: Fixed
https://github.com/yahoojapan/authorization-proxy/pull/66/commits/93d1112f058a87d924c1ef7b46faeabe30e1ef6d

**Pull request merged**